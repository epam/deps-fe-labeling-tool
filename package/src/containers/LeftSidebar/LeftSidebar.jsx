import React, { useState, useCallback, useRef, useEffect } from 'react'
import { DoubleLeftOutlined, DoubleRightOutlined } from '@ant-design/icons'
import PropTypes from 'prop-types'
import { Tabs } from '@/components/Tabs'
import { PreviewJson } from '@/containers/PreviewJson'
import { AllMarkupObjects } from '../AllMarkupObjects'
import {
  Sider,
  Toggler,
  TabsStyled,
  TabContent
} from './LeftSidebar.styles'

const { TabPane } = Tabs

const RESOURCE_TAB_JSON_PREVIEW = 'JSON Preview'
const RESOURCE_TAB_ALL_OBJECTS = 'All Objects'

const RESOURCE_TAB_TO_RENDER_CONTENT = {
  [RESOURCE_TAB_ALL_OBJECTS]: () => <AllMarkupObjects />,
  [RESOURCE_TAB_JSON_PREVIEW]: () => <PreviewJson />
}

const DEFAULT_ACTIVE_PAGE = '1'

const WIDTH_LEFT_SIDER = 300
const ZERO_WIDTH = 0
const TOGGLER_HEIGHT = 48

const LeftSidebarToggler = ({ onClick, collapsed, height }) => (
  <Toggler
    onClick={onClick}
    collapsed={collapsed}
    height={height}
  >
    {collapsed ? <DoubleRightOutlined /> : <DoubleLeftOutlined /> }
  </Toggler>
)

const LeftSidebar = ({ height }) => {
  const [collapsed, setCollapsed] = useState(true)
  const [contentHeight, setContentHeight] = useState(0)
  const contentWrapper = useRef()

  const onClick = useCallback(() => {
    setCollapsed((prevCollapsed) => !prevCollapsed)
  }, [])

  useEffect(() => {
    if (collapsed) {
      return
    }
    const tabsNavHeight = contentWrapper.current?.offsetTop ?? 0
    const tabContentHeight = height - tabsNavHeight - TOGGLER_HEIGHT
    setContentHeight(tabContentHeight)
  }, [height, collapsed])

  return (
    <Sider
      collapsed={collapsed}
      trigger={
        (
          <LeftSidebarToggler
            onClick={onClick}
            collapsed={collapsed}
            height={TOGGLER_HEIGHT}
          />
        )
      }
      collapsedWidth={ZERO_WIDTH}
      width={WIDTH_LEFT_SIDER}
      collapsible
    >
      <TabsStyled defaultActiveKey={DEFAULT_ACTIVE_PAGE} >
        {
          Object.keys(RESOURCE_TAB_TO_RENDER_CONTENT).map((item) => {
            return (
              <TabPane
                tab={item}
                key={item}
              >
                <TabContent
                  ref={contentWrapper}
                  height={contentHeight}
                >
                  {RESOURCE_TAB_TO_RENDER_CONTENT[item]()}
                </TabContent>
              </TabPane>
            )
          })
        }
      </TabsStyled>
    </Sider>
  )
}

LeftSidebarToggler.propTypes = {
  onClick: PropTypes.func.isRequired,
  collapsed: PropTypes.bool.isRequired,
  height: PropTypes.number
}

LeftSidebar.propTypes = {
  height: PropTypes.number
}

export {
  LeftSidebar
}
