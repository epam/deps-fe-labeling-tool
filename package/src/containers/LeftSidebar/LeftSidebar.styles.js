import styled, { css } from 'styled-components'
import { Sider } from '@/components/Sider'
import { Tabs } from '@/components/Tabs'
import { COLORS } from '@/theme/theme.default'

const mapPxToRem = (pxValue) => `${pxValue / 10}rem`

const StyledSider = styled(Sider)`
  z-index: 3;
  box-shadow: 1px 4px 14px rgba(0, 0, 0, 0.11);
  position: absolute;
  height: 100%;

  & .ant-layout-sider-zero-width-trigger {
    position: relative;
  }

  & .ant-collapse-header {
    min-height: 3.2rem;
    align-items: center !important;
    padding: 0 !important;
  }

  & .ant-tabs-nav-wrap {
    justify-content: center;
  }
`

const Toggler = styled.div`
  position: fixed;
  bottom: 0;
  left: 0;
  display: grid;
  align-items: center;
  align-self: end;
  justify-items: end;
  cursor: pointer;
  width: 30rem;
  box-shadow: 0 -1px 0 ${COLORS.GRAYSCALE_1};
  padding: 0 2rem;
  min-height: 4.8rem;
  background: ${COLORS.PRIMARY_3};

  ${(props) => props.collapsed && css`
    justify-items: center;
    border: 0.6px solid ${COLORS.GRAYSCALE_1};
    width: 3rem;
    height: 3rem;
    min-height: 0;
    box-shadow: none;
    bottom: 1.6rem;
    left: 1.6rem;
    padding: 0;
  `}
`

const TabsStyled = styled(Tabs)`
  .ant-tabs-tab {
    margin: 0.8rem 1.8rem !important;
  }
`

const TabContent = styled.div`
  overflow: auto;

  ${(props) => props.height && css`
    height: ${mapPxToRem(props.height)};
  `}
`

export {
  Toggler,
  StyledSider as Sider,
  TabsStyled,
  TabContent
}
