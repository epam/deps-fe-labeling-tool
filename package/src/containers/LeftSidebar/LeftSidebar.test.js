import React from 'react'
import { shallow } from 'enzyme'
import { LeftSidebar } from '@/containers/LeftSidebar'

describe('Container: LeftSidebar', () => {
  let defaultProps, wrapper

  beforeEach(() => {
    defaultProps = {
      height: 123
    }

    wrapper = shallow(<LeftSidebar {...defaultProps} />)
  })

  it('should render correct layout', () => {
    expect(wrapper).toMatchSnapshot()
  })
})
