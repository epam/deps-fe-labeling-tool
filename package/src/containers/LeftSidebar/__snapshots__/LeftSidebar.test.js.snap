// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Container: LeftSidebar should render correct layout 1`] = `
<LeftSidebarstyles__StyledSider
  collapsed={true}
  collapsedWidth={0}
  collapsible={true}
  trigger={
    <LeftSidebarToggler
      collapsed={true}
      height={48}
      onClick={[Function]}
    />
  }
  width={300}
>
  <LeftSidebarstyles__TabsStyled
    defaultActiveKey="1"
  >
    <TabPane
      key="All Objects"
      tab="All Objects"
    >
      <LeftSidebarstyles__TabContent
        height={0}
      >
        <Connect(AllMarkupObjects) />
      </LeftSidebarstyles__TabContent>
    </TabPane>
    <TabPane
      key="JSON Preview"
      tab="JSON Preview"
    >
      <LeftSidebarstyles__TabContent
        height={0}
      >
        <Connect(PreviewJson) />
      </LeftSidebarstyles__TabContent>
    </TabPane>
  </LeftSidebarstyles__TabsStyled>
</LeftSidebarstyles__StyledSider>
`;
